Prefix Compression and faster startup time

joka921 opened the pull request:
Current status: I am going to start a self-review later/tomorrow
This is a lot of change for one PR but they are separated relatively well from each other.

- Implemented the Prefix Compression Heuristic using a Trie-based
  greedy algorithm

- Integrated Prefix compression into the Vocabulary class
  (templated for prefix compression / no prefix compression)

- Converted SPO and SOP to Mmap based data

- faster startup time for ServerMain by not iterating over the complete
  Mmap Vector for statistics

- Added nlohnmann/json as a submodule
- IndexBuilder writes a configuration file (json) that contains
  information needed for the prefix compression. Can be extended for
  other settings and statistics

- it is also possible to externalize entities from the vocabulary (e.g.
  Wikidata statement ids)

- it is possible to pass a settings-json file to the index builder.
  Currently supported: Prefixes that shall be externalized (see above)

joka921 commented on src/MetaDataConverterMain.cpp:
```diff
@@ -37,7 +37,12 @@ int main (int argc, char** argv) {
+    /*convertHmapBasedPermutatationToMmap(permutName, permutName + ".converted",
                                         permutName + MMAP_FILE_SUFFIX);
+                                        */
+    addBlockListToMmapMetaDataPermutation(permutName,
```
Here I will add the automatic version detection so that everything is correctly converted and only if  necessary

niklas88 commented on src/index/Index.Text.cpp:
```diff
@@ -509,7 +509,7 @@ void Index::openTextFileHandle() {
 
 // _____________________________________________________________________________
-string Index::wordIdToString(Id id) const { return _textVocab[id]; }
+const string& Index::wordIdToString(Id id) const { return _textVocab[id]; }
```
Hmm, how does this not return a reference to a local variable? When the string is compressed, `AccessReturnType` is `std::string` and the result of `expandPrefix()`. Now this returns a `const string&` to it, but then that returned string immediately goes out of scope?!

joka921 replied:
The `_textVocab` is not compressed and thus is able to return a valid reference.
I could also change this return value to be the proper AccessReturnType but I liked this interface more and as soon as we would  think about compressing the text vocabulary, the compiler will complain.

niklas88 commented on src/index/IndexMetaData.h:
```diff
@@ -163,6 +169,13 @@ using IndexMetaDataMmapView = IndexMetaData<
 // constants for Magic Numbers to separate different types of MetaData;
 const size_t MAGIC_NUMBER_MMAP_META_DATA = static_cast<size_t>(-1);
 const size_t MAGIC_NUMBER_SPARSE_META_DATA = static_cast<size_t>(-2);
-const size_t MAGIC_NUMBER_MMAP_META_DATA_BLOCK_LIST = static_cast<size_t>(-3);
```
where is this gone? The version is still called `V_BLOCK_LIST_AND_STATISTICS`

joka921 replied:
Yes, I would suggest a magic number that stays fixed from now on and a version Tag which can be incremented. The magic number must still discriminate whether we have a version tag at all, and thus we have 4 Magic Numbers for sparse/dense and with/without version.

The version tag describes the features of this version (thus "block list" and "statistics").

In one of the next PRs I will clean the serialization/deserialization of the IndexMetaData a bit more, e.g. only have one createFromByteBuffer that handles sparse and dense meta data using constexpr if etc.

niklas88 replied:
I thought wherever we added a magic number we also added a version field? If not maybe we should just drop support for those files where we didn't do that?

joka921 replied:
Okay, this is a design decision:
1. Dropping the support makes the code easier
2. Dropping the support breaks ALL Indices created so far
3. We could also archive the old versions of createFromByteBuf somewhere for the MetaDataConverter. We would then have some code duplications, but somewhere in a frozen state far away from the actual production code.

Just tell me what you like best.

niklas88 replied:
I mean didn't we add a version wherever we added a magic number? So that we only got two formats
* No magic number and no version = legacy format
* Magic number and version = can increment version and keep magic number

joka921 replied:
Unfortunately we forgot to add a version with the IndexMetaData.
But I think we could implement your suggestion and we would break only the Indices built in the last days, using the "MMapBasedVector"-PR. I think this only affects your freebase builds which were converted from old builds and could be converted again. Then we would only have ONE magic number for sparse/dense each and the rest is handled by a version tag which I consider to be cleaner. What do you think of this?

niklas88 replied:
I converted most of our QLever indices already and @hannahbast  built a new version of Wikipedia + Freebase Easy but that's ok. I'd rather we break some intermediate indices now than be left with a worse design later.

joka921 replied:
Ok, I think I finally got you now.( I first implemented a version that can read ANYTHING). What I can do with this design is Inform the User, that his index can not be fixed (should only affect you and Prof.Bast)

joka921 commented on .gitmodules:
```diff
@@ -4,3 +4,6 @@
 [submodule "third_party/googletest"]
 	path = third_party/googletest
 	url = https://github.com/google/googletest.git
+[submodule "third_party/json"]
```
Yes indeed, how do I choose/fixate a certain commit of a certain branch of the submodule for my project?

niklas88 replied:
https://stackoverflow.com/a/5828396/692303

joka921 commented on CMakeLists.txt:
```diff
@@ -106,6 +113,9 @@ target_link_libraries (WriteIndexListsMain engine ${CMAKE_THREAD_LIBS_INIT})
 add_executable(MetaDataConverterMain src/MetaDataConverterMain.cpp)
 target_link_libraries (MetaDataConverterMain metaConverter ${CMAKE_THREAD_LIBS_INIT})
 
+add_executable(PrefixHeuristicMain src/PrefixHeuristicMain.cpp)
```
Yes, that is right. In fact this currently does not do anything "useful" for  Qlever but I probably can use it for my thesis to evaluate the potential of the method. I integrated it here to make use of the easy linking with cmake. Is it okay to leave it there? 

And actually you remind me of something: we have to update the converter or some part of the program to also convert/compress the vocabulary to the new format (one of the disadvantages of "the first char is always a prefix code"

niklas88 replied:
I think we should keep it but it should say what it does

joka921 commented on e2e/e2e.sh:
```diff
@@ -41,7 +41,7 @@ INDEX="e2e_data/scientists-index"
 	rm -f "$INDEX.*"
 	pushd "./build"
-	./IndexBuilderMain -a -l -i "../$INDEX" \
+	./IndexBuilderMain -a -l -c -i "../$INDEX" \
```
If you agree to make compression the default (which perfectly makes sense), I happily agree.

niklas88 replied:
Yes I think we should do that, especially since it's a different vocabulary format and otherwise that will lead to confusion.

joka921 commented on src/index/Vocabulary.h:
```diff
@@ -167,42 +249,139 @@ class Vocabulary {
+  //            in the same order as the infile
+  //   prefixes - a list of prefixes which we will compress
+  //
+  //   Returns: A json array with information about the prefixes,
```
Yes you are right, then we can handle the internals of the .configuration file all in one place in the index class.

joka921 replied:
I changed it for this place. But I am not so sure anymore if this is the way to go (it also affects some different places):
This is a serialization information, which the Vocabulary needs exactly in this form to setup the compression when initializing an index from disk. The index class does not need to  understand the  exact  format but only has to store this serialized information (in this case a json type) in a file.

I think this is a common pattern when serializing, just call serialize on all the members:)

joka921 commented:
And don't worry about that test failure, I found the bug and will fix it tomorrow afternoon.
The test is not good there: it implicitly checks implementation details (assumes a certain size for the serialization of the MetaData instead of explicitly checking it or blackbox test it). The effect was a nondeterministic test failure (works on my machine) for code which is probably fine. You'll see what I mean tomorrow.

niklas88 commented:
I think you check them out locally and commit that explicitly and then everyone else needs to do `git submodule sync`. I'm not 100% sure though.On Aug 22, 2018 18:02, joka921 <notifications@github.com> wrote:@joka921 commented on this pull request.



In .gitmodules:
> @@ -4,3 +4,6 @@
 [submodule "third_party/googletest"]
 	path = third_party/googletest
 	url = https://github.com/google/googletest.git
+[submodule "third_party/json"]

Yes indeed, how do I choose/fixate a certain commit of a certain branch of the submodule for my project?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

niklas88 commented on src/index/MetaDataConverter.h:
```diff
@@ -41,12 +41,36 @@ void convertPermutationToHmap(const string& permutIn, const string& permutOut,
+// __________________________________________________________________________
+inline void notifyCreated(const string& filename, bool hasConvertedSuffix) {
+  if (hasConvertedSuffix) {
+    std::cout << "created new file " << filename
```
This is nice

niklas88 commented on src/index/VocabularyImpl.h:
```diff
@@ -0,0 +1,347 @@
+template <typename>
+string Vocabulary<S>::expandPrefix(const CompressedString& word) const {
+  assert(!word.empty());
+  auto idx = static_cast<unsigned char>(word[0]) - MIN_COMPRESSION_PREFIX;
```
I find `unint8_t` clearer here because it's not really a character

niklas88 commented on src/index/VocabularyImpl.h:
```diff
@@ -0,0 +1,347 @@
+  if (idx < NUM_COMPRESSION_PREFIXES) {
+    return _prefixMap[idx] + word.toStringView().substr(1);
+  } else {
+    return string(word.toStringView().substr(1));
```
I think the outer `string()` is redundant here

joka921 replied:
No, it is not. Converting from string_view to string is not possible and construction is explicit.

niklas88 commented on src/index/VocabularyImpl.h:
```diff
@@ -0,0 +1,347 @@
+    if (ad_utility::startsWith(word, p._fulltext)) {
+      auto res = CompressedString::fromString(
+          p._prefix + std::string_view(word).substr(p._fulltext.size()));
+      LOG(DEBUG) << "compressed " << word << " to " << res.toString() << '\n';
```
is that `toString()` needed? I'd think that `ostream` handles `string_view`?

joka921 replied:
This is not a string view but a CompressedString. But I can teach ostream to do this, would probably be cleaner. But in this place I anyway want to throw out that LOG, because it is unreadable and much too verbose for every word. (This was for tracking a bug, that has been removed for some time now).

niklas88 commented on src/index/VocabularyImpl.h:
```diff
@@ -0,0 +1,347 @@
+    el = "";
+  }
+  _prefixVec.clear();
+  unsigned char idx = 0;
```
`unit8_t`

joka921 commented on src/index/CompressedString.h:
```diff
@@ -0,0 +1,60 @@
+  }
+
+  // explicit conversions to strings and string_views
+  string toString() const { return *this; }
```
c++ stl has std::to_string(42); But if this is the convention I am going to change it.

niklas88 replied:
No makes sense to keep it in our camel case format.

joka921 commented on src/index/Index.Text.cpp:
```diff
@@ -167,6 +172,10 @@ void Index::passContextFileIntoVector(const string& contextFile,
       bool ret = _textVocab.getId(line._word, &wid);
+      if (!ret) {
+        LOG(INFO) << "ERROR: word " << line._word
+                  << "not found in textVocab. Terminating\n";
```
I am not sure I understand you. When we are not in DEBUG mode, the return value is not even checked which as faster and strange things are going to happen when this fails. I think @Buchhold implemented some of those to allow fast creation of indices, but better error handling in Debug mode.
The extra LOG was inserted by me to track down an error (no good idea to to use `\n` as a compression prefix) but I think it's nice to keep it. Then, when debugging we not only have a failed assertion but immediately an hint where to look.

joka921 commented on src/index/Index.cpp:
```diff
@@ -1622,3 +1645,72 @@ void Index::setKeepTempFiles(bool keepTempFiles) {
+    _onDiskLiterals = _configurationJson["external-literals"];
+  }
+
+  if (_configurationJson.find("prefixes") != _configurationJson.end()) {
```
Well, I think my whole wording was misleading. Actually the `.configuration` or `.conf` file is not to be changed at all by the user. It is there to transfer required information from the IndexBuilder to the Server. (currently "which prefixes do compress" "which prefixes are externalized" "do we have external literals"). All three are required to be correct/ unchanged in order to make the server work. I could add a preamble to this file that it must not be changed. My thoughts were the following
1. JSON is easy to read, so we can easily see what the index does.
2. JSON is a serialization format for which you accepted including external libraries. Currently we are doing all serialization manually (Index Meta Data, Vocabulary, etc) which is not always easy to maintain. Prepending this to the vocabulary would be easier but could we think about integrating a lightweight serialization library, that can also serialize classes to binary? Then we could only change one place when changing something for the read/write of types.
3. If we want to keep dependencies small, we could write a simple custom ad_utility::serializer to make this whole `f.write(&something, sizeof(something)` and somewhere completely else `f.read(&something, sizeof(something)` a bit better readable and better maintainable.
3. With Json we can easily serialize different types.

joka921 commented on src/index/IndexBuilderMain.cpp:
```diff
@@ -38,6 +39,8 @@ struct option options[] = {{"all-permutations", no_argument, NULL, 'a'},
                            {"words-by-contexts", required_argument, NULL, 'w'},
                            {"add-text-index", no_argument, NULL, 'A'},
                            {"keep-temporary-files", no_argument, NULL, 'k'},
+                           {"settings-file", required_argument, NULL, 's'},
```
This is not the .conf file (see above). This is meant to be a file where we can actually specify something:)

niklas88 commented on src/global/Constants.h:
```diff
@@ -51,12 +51,16 @@ static const int DEFAULT_NOF_VALUE_MANTISSA_DIGITS = 30;
-static constexpr size_t MIN_COMPRESSION_PREFIX = 128;
-static constexpr size_t NUM_COMPRESSION_PREFIXES = 127;
+// Constants for the range of valid compression prefixes
+// all printable characters are left out
```
Say ASCII not printable, we are assuming UTF-8 in most of QLever. Also, is this really relevant now that we always use the first byte to indicate compression. We could just as easily use `'N'` and `'C'` for (non-) compressed.

niklas88 commented:
Can you rebase on the freshly merged disambiguate optional work with the changes to `operator[]` as discussed in that PR. After that I think we can merge this.

niklas88 commented on src/PrefixHeuristicEvaluatorMain.cpp:
```diff
@@ -34,7 +34,7 @@ int main(int argc, char** argv) {
   }
 
-  for (const auto& p : calculatePrefixes(argv[1], 127, 1)) {
+  for (const auto& p : calculatePrefixes(argv[1], 127, 1, true)) {
```
I think this should use the `NUM_COMPRESSION_PREFIXES` constant so that it always matches the encoding actually in use

niklas88 commented:
So am I seeing this right, the last couple of e2e test failures were just flakyness?

joka921 commented:
@niklas88 The first one was a bug with getopt (I removed the -l flag which was still present in the e2e test).
The second failure I did not understand (worked fine locally, also with Release builds). Obviously the server failed to open, so I added a Server run in the foreground which of course let the test stall but yielded no results. After simply removing this again it worked fine so I assume that this one build I cannot explain (3ccb9ec) was some versioning/checkout problem with Travis.

joka921 commented:
I Implemented the changes we discussed on Monday so I think we are good for now. The things that could be improved are all stuff for separate PRs in my opinion. If you agree I will squash the commits and merge this, I just wanted to give you the chance to look at it again first.
